Add documentation for starting and exiting (#12645)
nvaccess/nvda: devDocs/startupShutdownInstall.md

# NVDA Starting and Exiting outline

## Ways to start NVDA:

1. For an installed copy:
    1. Ctrl+Alt+N (Desktop shortcut)
    1. Automatically via Ease of Access on the Windows sign-in screen (at boot or signing out of a previous session)
    1. Automatically via Ease of Access on User Account Control (UAC) screens
    1. Automatically by Ease of Access after signing in to Windows
1. For an installed copy, portable copy, installer:
    1. An exiting instance of NVDA starting a new process (see shutting down procedures)
    1. By running the exe (see cli options too).
    This can be triggered by a user or external process such as an existing NVDA instance
    - test: `startupShutdownNVDA.Starts`

> 1. For source: eg runnvda.bat

Reviewer: I think this is a good overview. I'd also like a "test plan" for each of these (perhaps adding a new heading in a new "test plan" section for each of these. Test plans are particularly useful for demonstrating the edge cases around updating, and providing the detailed steps and expectations for a manual tester. If we have to do manual testing, this can save a lot of time by spelling out how to test, especially for the trickier use cases. Ideally when someone is modifying this area of NVDA we can say follow this test plan to confirm everything still works.

1. UI within NVDA, with and without an ExitDialog prompt (uses `triggerNVDAExit`):
    1. NVDA+q
        - test: `startupShutdownNVDA.Quits from keyboard, Restarts`
    1. An input gesture to restart
    1. A prompt to restart from changing certain settings
    1. Via the NVDA menu -> Exit
        - test: `startupShutdownNVDA.Quits from menu`
1. A process sending `WM_QUIT`, eg a new NVDA process starting (uses `WM_QUIT`)
1. A handled crash (directly causes a new process to start, terminates unsafely)
    - test: `startupShutdownNVDA.Restarts on crash, Restarts on braille crash`
1. An unhandled crash (terminates unsafely)
1. An external command which kills the process (terminates unsafely) 
1. Windows shutting down (terminates unsafely) (uses `wx.EVT_END_SESSION`)

## Exit hooks/triggers

### In process: `triggerNVDAExit`
* A function in the core module
* Only executes the code once, uses a lock and flag to ensure this
* Uses a queue on the main thread to queue a safe shutdown
* Once the queued shutdown starts:
    1. the updateCheck is terminated
    1. watchdog is terminated
    1. addons and the brailleViewer are terminated, so we can close all windows safely

>     1. All NVDA windows are closed

Author: ```suggestion
    1. All wx windows are closed
```

    1. Now that windows are closed, a new NVDA instance is started if requested

### Out of process: `WM_QUIT`
* [a windows message](https://docs.microsoft.com/en-us/windows/win32/winmsg/wm-quit) which can be sent across processes which will force the main loop to exit and close most NVDA windows.
* We subsequently run `triggerNVDAExit` to ensure that clean up code isn't missed, and pump the queue to execute it.
* Changing to using a custom message, which would allow a custom handling (eg just `triggerNVDAExit`), requires compatibility with messaging older NVDA versions that are only aware of `WM_QUIT`.

### Out of process: `wx.EVT_END_SESSION`
- This is a [wxCloseEvent](https://docs.wxwidgets.org/3.0/classwx_close_event.html) triggered by a Windows session ending.
- On `wx.EVT_END_SESSION`, we save the config and play the exit sound.

> - Other actions are not performed as we have limited time to perform an action for this event. With NVDA being killed off very last, Windows may include NVDA in the Block shutdown dialog, but the user won't be able to read it if we are shutting down.

Author: This a hook for exiting NVDA, but the separation between these sections is pretty meaningless (other than formatting)

## Technical startup and exit notes

## The Message Window

> The Windows API allows us to find a process by searching for a named window. To make the NVDA process searchable, NVDA creates a message window with a specific name.

Reviewer: > named window

This is talking about 'window class' right? If so, perhaps "by searching for a named window (window class)"

Author: no, we use the standard window class name, we just give it a custom name attribute.

Author: Just a correction, we use the (old) standard wxWindow class name, `wxWindowClassNR`

Author: Microsoft generally just calls these objects "Windows" in the documentation, which makes things confusing.

> Starting up is blocked by checking for an existing NVDA window and sending `WM_QUIT`.

Reviewer: ```suggestion
Starting up is blocked by checking for an existing NVDA window (via the a known Window Class value) and sending `WM_QUIT`.

```

Author: ```suggestion
Starting up is blocked by checking for an existing NVDA window (via a named custom Window) and sending `WM_QUIT`.
```

We then set the NVDA process to allow the `WM_QUIT` to be sent from other processes (see [WM_QUIT](#Out-of-process-WM_QUIT)). 

The message window is created late during the start up, and destroyed early in exit and is not perfectly indicative of whether or not an NVDA process is running. As such, we have a [MutEx](#MutEx) that is checked as well.

> 

Reviewer: I think the intent of the message window / window class lookup, is so that that `WM_QUIT` can be sent. Otherwise we can't identify the process, right? This only matters if NVDA isn't already exiting. If the message window doesn't exist, it is either too early, or NVDA is already exiting.

I'd try to make this intent clear:
- The WM_QUIT / window class mechanism is to make it possible to trigger another NVDA process to START to exit.
- The MutEx ensures that a newly started process blocks until any previous NVDA has FINISHED exiting.

Author: I've rewritten this to make the purpose of the MW clearer. The Message Window is used for handling other windows messages (see `MessageWindow`), not just `WM_QUIT`. But in regards to this document, that is why it is relevant.

# NVDA Starting and Exiting outline

## Ways to start NVDA:

1. For an installed copy:
    1. Ctrl+Alt+N (Desktop shortcut)
    1. Automatically via Ease of Access on the Windows sign-in screen (at boot or signing out of a previous session)
    1. Automatically via Ease of Access on User Account Control (UAC) screens
    1. Automatically by Ease of Access after signing in to Windows

>     1. An exiting instance of NVDA starting a new process (see shutting down procedures)
> 1. For an installed copy, portable copy, installer:
>     - By running the exe (see cli options too).

Author: ```suggestion
1. For an installed copy, portable copy, installer:
    1. An exiting instance of NVDA starting a new process (see shutting down procedures)
    1. By running the exe (see cli options too).
```

    This can be triggered by a user or external process such as an existing NVDA instance
    - test: `startupShutdownNVDA.Starts`
1. For source: eg runnvda.bat

## NVDA can be shutdown by:

1. UI within NVDA, with and without an ExitDialog prompt (uses `triggerNVDAExit`):
    1. NVDA+q
        - test: `startupShutdownNVDA.Quits from keyboard, Restarts`

>     1. An unbound command to restart

Reviewer: What does this mean? Is it referring to a command that by default has no gesture bound to it?

>     1. A prompt to restart from changing certain settings

Reviewer: I think this can be grouped with "via the UI" (eg with NVDA+q, or via the exit option in the menu)

Author: I think this is currently the case, unless you meant something different

    - test: `startupShutdownNVDA.Restarts on crash, Restarts on braille crash`
1. An unhandled crash (terminates unsafely)
1. An external command which kills the process (terminates unsafely) 
1. Windows shutting down (terminates unsafely)
    - On `wx.EVT_END_SESSION`, we save the config and play the exit sound. (Note: should this do more? NVDA may terminate unsafely directly after this)

## Safe exit hooks/triggers

### In process: `triggerNVDAExit`
* A function in the core module
* Only executes the code once, uses a lock and flag to ensure this
* Uses a queue on the main thread to queue a safe shutdown
* Once the queued shutdown starts:
    1. the updateCheck is terminated
    1. watchdog is terminated
    1. addons and the brailleViewer are terminated, so we can close all windows safely
    1. All NVDA windows are closed
    1. Now that windows are closed, a new NVDA instance is started if requested

### Out of process: `WM_QUIT`
* a windows message which can be sent across processes which will force the main loop to exit and close most NVDA windows.
* We subsequently run `triggerNVDAExit` to ensure that clean up code isn't missed, and pump the queue to execute it.

> ## Technical startup and exit process

Reviewer: I don't think this section is useful.

# NVDA Starting and Exiting outline

## As a user, NVDA can be started by:

1. For an installed copy:

>     1. Ctrl+Alt+N

Reviewer 2: ```suggestion
    1. Ctrl+Alt+N (Desktop shortcut)
```

>     1. Via secure desktop/ease of access

Reviewer 2: ```suggestion
    1. Automatically via Ease of Access on the Windows sign-in screen (at boot or signing out of a previous session)
    1. Automatically via Ease of Access on User Account Control (UAC) screens
```

>     1. Via startup

Reviewer 2: ```suggestion
    1. Automatically by Ease of Access after signing in to Windows
    1. An automatic restart of NVDA due to a handled crash
```

Author: >    1. An automatic restart of NVDA due to a handled crash

Should all the ways of restarting NVDA be mentioned here?

Maybe just 

```suggestion
    1. Automatically by Ease of Access after signing in to Windows
    1. An exiting instance of NVDA starting a new process (see shutting down procedures)
```

Reviewer 2: That would be fine with me.

1. For an installed copy, portable copy, installer:
    - By running the exe (see cli options too).
    This can be triggered by a user or external process such as an existing NVDA instance
    - test: `startupShutdownNVDA.Starts`
1. For source: eg runnvda.bat

> ## As a user, NVDA can be shutdown by:

Reviewer: I think the "as a user" part can be removed from here and from the previous heading. Some of these aren't direct user actions and I'm not sure it is a useful distinction.

I suppose there are two "views" of shutdown / startup that are useful:
- What can I "a tester" do to exercise all the startup / shutdown pathways?
  - Normal shutdown
     - Trigger via one of (NVDA+q, NVDA+n and select exit)
     - System tests
        - "startupShutdownNVDA.Quits from menu"
  - Etc
- What do I "a developer" need to be aware of for startup / shutdown mechanisms"
   -  Normal shutdown (most common):
      - UI calls `triggerNVDAExit`
      - All wx windows close (Note, "NVDA window class" no longer present)
      - wx event loop ends
      - etc
   - receive WM_QUIT:
      - wx Mainloop exits
      - triggerNVDAExit is called
   - restarting due to UI
      - core.restart called, this calls core.triggerNVDAExit
   - restarting due to crash
      - core.restartUnsafely called 
      - Nothing can be done about current instance, just start a new instance.

Author: ```suggestion
## NVDA can be shutdown by:
```

Author: Yep, these were the two views I had. Just phrasing the titles was awkward.

1. UI within NVDA, with and without an ExitDialog prompt (uses `triggerNVDAExit`):
    1. NVDA+q
        - test: `startupShutdownNVDA.Quits from keyboard, Restarts`
    1. An unbound command to restart
    1. A prompt to restart from changing certain settings
    1. Via the NVDA menu -> Exit
        - test: `startupShutdownNVDA.Quits from menu`
1. A process sending `WM_QUIT`, eg a new NVDA process starting (uses `WM_QUIT`)
1. A handled crash (directly causes a new process to start, terminates unsafely)
    - test: `startupShutdownNVDA.Restarts on crash, Restarts on braille crash`
1. An unhandled crash (terminates unsafely)
1. An external command which kills the process (terminates unsafely) 
1. Windows shutting down (terminates unsafely)

>     - On `wx.EVT_END_SESSION`, we save the config and play the exit sound. (Note: should this do more? NVDA may terminate unsafely directly after this)

Reviewer 2: Windows does not give us very long to exit here. If we take too long we will be killed off anyway I believe. So saving config is the most important thing we should do. 
See https://docs.microsoft.com/en-us/windows/win32/shutdown/wm-endsession

Author: It seems like the urgency depends on the parameters passed in (eg the context of the shutdown). In my experience, apps can block a regular shutdown for as long they wish (Google Drive takes a while for me sometimes, unsaved notepads block indefinitely).

Reviewer 2: The only difference for us is that in nvda.pyw we call `SetProcessShutdownParameters` which tells Windows to kill us off after all other apps. The reason we do this is so that if other apps take time to close, we can still actually as a screen reader read the Blocked Shutdown dialog and the list of apps, so that the NVDA user can accessibly go and manually shutdown those apps / answer any save dialogs etc. So, then with NVDA being killed off very last, if we hourselves seem to take too much time by Windows, then we ourselves may get included in the Block shutdown dialog, but the user won't be able to read it most probobally, as we would already be somewhere within our shutdown sequence.

Author: Is it worth adding this information to the docs?

```suggestion
    - On `wx.EVT_END_SESSION`, we save the config and play the exit sound.
        - Other actions are not performed as we have limited time to perform an action for this event. With NVDA being killed off very last, Windows may include NVDA in the Block shutdown dialog, but the user won't be able to read it if we are shutting down.
```

## Safe exit hooks/triggers

### In process: `triggerNVDAExit`
* A function in the core module
* Only executes the code once, uses a lock and flag to ensure this
* Uses a queue on the main thread to queue a safe shutdown
* Once the queued shutdown starts:
    1. the updateCheck is terminated
    1. watchdog is terminated
    1. addons and the brailleViewer are terminated, so we can close all windows safely
    1. All NVDA windows are closed
    1. Now that windows are closed, a new NVDA instance is started if requested

### Out of process: `WM_QUIT`
* a windows message which can be sent across processes which will force the main loop to exit and close most NVDA windows.

> * We subsequently run `triggerNVDAExit` to exit other required processes, and pump the queue to execute the shutdown.

Reviewer 2: What does "other required processes" mean? Do you mean other copies of NVDA, or do you mean other subcomponants of NVDA?

Author: ```suggestion
* We subsequently run `triggerNVDAExit` to ensure that clean up code isn't missed, and pump the queue to execute it.
```

Author: I just mean what `triggerNVDAExit` does, which I was trying to avoid from repeating.

Reviewer 2: This reads better to me now, thanks.

## Technical startup and exit process

All the startup pathways execute nvda.pyw.

1. *Start nvda.pyw* 
1. MonkeyPatches are applied
1. The virtual environment is checked
1. Gettext is installed
1. Starting arguments are parsed
1. We check for a running instance of NVDA by looking for a Message Window
    - terminate it if possible using `WM_QUIT` or forcing the process to exit. If this fails, abandon startup

> 1. We try to acqure the NVDA mutex

Reviewer 2: ```suggestion
1. We try to acquire the NVDA mutex
```

    - abandon startup if we can't
1. Initialise logging
1. Set the process to allow `WM_QUIT` messages
1. Start running `core.main()`
    1. SetProcessDPIAware
    1. Initialize config

>     1. Play statup sound

Reviewer 2: ```suggestion
    1. Play startup sound

```

    1. Set language
    1. Initialize NVDA modules and wx GUI

>     1. Start the Message Window (used for `WM_QUIT`)

Reviewer 2: ```suggestion
    1. Create the Message Window (used for `WM_QUIT`)
```

>     1. Start the core pump

Reviewer 2: ```suggestion
    1. Start the core pump (the timer that is fired inside the main loop when there is work to do)
```

    1. Start watchdog
    1. Notify postNvdaStartup
    1. Start the wx app main loop
    --- 
    1. *The main loop has exited via `triggerNVDAExit` or `WM_QUIT`*
    1. If `WM_QUIT` was used to exit the main loop, perform `triggerNVDAExit`
    1. Terminate gui
    1. Save config
    1. Terminate NVDA modules
    1. Play exit sound
    1. Destroy the message window
    1. Return to nvda.pyw
1. Notify ease of access of the shutdown
1. Release the Mutex

> 1. Exit nvda.pyw with `sys.exit`

Reviewer: I'd avoid this level of detail. It will be hard to keep it in sync with the code. Instead I'd focus on the key points:
- NVDA protects itself from having multiple instances running:
  - By detecting the presence of a Window Class, in this case WM_QUIT is used to signal the old process.
  - By waiting to acquire an OS mutex.

Author: Which aspects do you think are unnecessary? I'm going to suggest grouping some but I do think these are all fairly core parts of the shutdown/start process

Reviewer: The code serves as a better reference. Keeping this accurate (in sync) with code will be difficult. This isn't necessarily highlighting information that isn't already in the code, low signal to noise ratio.

Instead I'd focus on key aspects of the code that may not be immediately obvious. If you can draw attention to facts about the way the code works, it becomes easier to reason about and debug when there are issues. For instance, after the mutex is acquired there can not be another copy of NVDA running in that "critical section" (which extends until the final moments of shutting down).

# NVDA Starting and Exiting outline

> ## As a user, NVDA can be started by:

Author: ```suggestion
## Ways to start NVDA:
```

1. An unhandled crash (terminates unsafely)
1. An external command which kills the process (terminates unsafely) 
1. Windows shutting down (terminates unsafely)
    - On `wx.EVT_END_SESSION`, we save the config and play the exit sound. (Note: should this do more? NVDA may terminate unsafely directly after this)

## Safe exit hooks/triggers

### In process: `triggerNVDAExit`
* A function in the core module
* Only executes the code once, uses a lock and flag to ensure this
* Uses a queue on the main thread to queue a safe shutdown
* Once the queued shutdown starts:
    1. the updateCheck is terminated
    1. watchdog is terminated
    1. addons and the brailleViewer are terminated, so we can close all windows safely
    1. All NVDA windows are closed
    1. Now that windows are closed, a new NVDA instance is started if requested

### Out of process: `WM_QUIT`
* a windows message which can be sent across processes which will force the main loop to exit and close most NVDA windows.
* We subsequently run `triggerNVDAExit` to exit other required processes, and pump the queue to execute the shutdown.

## Technical startup and exit process

All the startup pathways execute nvda.pyw.

> 1. *Start nvda.pyw* 
> 1. MonkeyPatches are applied
> 1. The virtual environment is checked
> 1. Gettext is installed
> 1. Starting arguments are parsed
> 1. We check for a running instance of NVDA by looking for a Message Window
>     - terminate it if possible using `WM_QUIT` or forcing the process to exit. If this fails, abandon startup
> 1. We try to acquire the NVDA mutex
>     - abandon startup if we can't
> 1. Initialise logging
> 1. Set the process to allow `WM_QUIT` messages
> 1. Start running `core.main()`
>     1. SetProcessDPIAware
>     1. Initialize config
>     1. Play startup sound
>     1. Set language
>     1. Initialize NVDA modules and wx GUI
>     1. Create the Message Window (used for `WM_QUIT`)
>     1. Start the core pump (the timer that is fired inside the main loop when there is work to do)
>     1. Start watchdog
>     1. Notify postNvdaStartup
>     1. Start the wx app main loop
>     --- 
>     1. *The main loop has exited via `triggerNVDAExit` or `WM_QUIT`*
>     1. If `WM_QUIT` was used to exit the main loop, perform `triggerNVDAExit`
>     1. Terminate gui
>     1. Save config
>     1. Terminate NVDA modules
>     1. Play exit sound
>     1. Destroy the message window
>     1. Return to nvda.pyw
> 1. Notify ease of access of the shutdown
> 1. Release the Mutex
> 1. Exit nvda.pyw with `sys.exit`

Author: ```suggestion
1. *Start nvda.pyw* 
1. MonkeyPatches are applied, the virtual environment is checked, starting arguments are parsed
1. We check for a running instance of NVDA by looking for a Message Window
    - terminate it if possible using `WM_QUIT` or forcing the process to exit. If this fails, abandon startup
1. We try to acquire the NVDA mutex
    - abandon startup if we can't
1. Initialise logging
1. Set the process to allow `WM_QUIT` messages
1. Start running `core.main()`
    1. Initialize config
    1. Play startup sound
    1. Set language, initialize NVDA modules and wx GUI
    1. Create the Message Window (used for `WM_QUIT`)
    1. Start the core pump (the timer that is fired inside the main loop when there is work to do)
    1. Start watchdog
    1. Notify postNvdaStartup
    1. Start the wx app main loop
    --- 
    1. *The main loop has exited via `triggerNVDAExit` or `WM_QUIT`*
    1. If `WM_QUIT` was used to exit the main loop, perform `triggerNVDAExit`
    1. Terminate gui
    1. Save config
    1. Terminate NVDA modules
    1. Play exit sound
    1. Destroy the message window
    1. Return to nvda.pyw
1. Notify ease of access of the shutdown
1. Release the Mutex
1. Exit nvda.pyw with `sys.exit`
```
